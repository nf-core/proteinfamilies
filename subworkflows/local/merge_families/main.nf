/*
    FAMILY MERGING
*/

include { POOL_SIMILAR_COMPONENTS } from '../../../modules/local/pool_similar_components/main'
include { MERGE_SEEDS             } from '../../../modules/local/merge_seeds/main'
include { GENERATE_FAMILIES       } from '../../../subworkflows/local/generate_families'

workflow MERGE_FAMILIES {
    take:
    similarities                     // tuple val(meta), path(txt)
    seed_msa                         // tuple val(meta), path(aln)
    sequences                        // tuple val(meta), path(fasta)
    alignment_tool                   // string ["famsa", "mafft"]
    skip_msa_trimming                // boolean
    clipkit_out_format               // string (default: clipkit)
    hmmsearch_write_target           // boolean
    hmmsearch_write_domain           // boolean
    recruit_sequences_with_models    // boolean
    hmmsearch_query_length_threshold // number [0.0, 1.0]

    main:
    ch_versions = Channel.empty()

    POOL_SIMILAR_COMPONENTS( similarities )
    ch_versions = ch_versions.mix( POOL_SIMILAR_COMPONENTS.out.versions )

    ch_pooled_components = POOL_SIMILAR_COMPONENTS.out.pooled_components
        .splitCsv( by:1 )
        .map { meta, components ->
            // Extract the suffix from each component (part after last underscore)
            def suffixes = components.collect { component ->
                component.split('_').last()
            }
            // Join the suffixes with underscores
            def combinedSuffix = suffixes.join('_')
            // Keep original id, add new field merged_id
            def newMeta = meta + [merged_id: "${meta.id}_${combinedSuffix}"]
            return [newMeta, components]
        }

    MERGE_SEEDS( ch_pooled_components, seed_msa.first() )
    ch_versions = ch_versions.mix( MERGE_SEEDS.out.versions )

    GENERATE_FAMILIES( sequences, MERGE_SEEDS.out.merged_seed_msa, \
            alignment_tool, skip_msa_trimming, clipkit_out_format, \
            hmmsearch_write_target, hmmsearch_write_domain, \
            recruit_sequences_with_models, hmmsearch_query_length_threshold )
    ch_versions = ch_versions.mix( GENERATE_FAMILIES.out.versions )

    emit:
    versions = ch_versions
    seed_msa = GENERATE_FAMILIES.out.seed_msa
    full_msa = GENERATE_FAMILIES.out.full_msa
    fasta    = GENERATE_FAMILIES.out.fasta
    hmm      = GENERATE_FAMILIES.out.hmm
}
