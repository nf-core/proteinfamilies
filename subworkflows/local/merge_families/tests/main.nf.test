nextflow_workflow {

    name "Test Subworkflow MERGE_FAMILIES"
    script "../main.nf"
    workflow "MERGE_FAMILIES"

    test("fasta") {

        when {
            workflow {
                """
                def similarities_content = "family_1,family_2,similarity_score\\nmgnifams_test_1,mgnifams_test_2,0.94\\n"

                def similarities_file = file('similarities.csv')
                similarities_file.text = similarities_content

                input[0] = [
                    [ id: 'test' ],
                    similarities_file
                ]
                input[1] = Channel.of([
                    [ id: 'test' ],
                    [
                        file(params.pipelines_testdata_base_path + 'test_data/modules/mgnifams_test_1.aln', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'test_data/modules/mgnifams_test_2.aln', checkIfExists: true)
                    ]
                ])
                input[2] = Channel.of([
                    [ id: 'test' ],
                    file(params.pipelines_testdata_base_path + 'test_data/mgnifams_input_small.fa', checkIfExists: true)
                ])
                input[3] = "famsa"   // alignment_tool
                input[4] = false     // skip_msa_trimming
                input[5] = "clipkit" // clipkit_out_format
                input[6] = false     // hmmsearch_write_target
                input[7] = true      // hmmsearch_write_domain
                input[8] = false     // skip_additional_sequence_recruiting
                input[9] = 0.9       // hmmsearch_query_length_threshold
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    file(workflow.out.hmm[0][1]).name,
                    workflow.out.seed_msa,
                    workflow.out.full_msa,
                    workflow.out.fasta,
                    workflow.out.versions.collect { path(it).yaml }.unique()
                    ).match()}
            )
        }
    }

}
