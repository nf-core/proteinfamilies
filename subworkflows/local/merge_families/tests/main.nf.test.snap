{
    "fasta": {
        "content": [
            "test_null.hmm.gz",
            [
                [
                    {
                        "id": "test",
                        "merged_id": "test_1_2"
                    },
                    "test_null.clipkit:md5,28f6460dd199c925d44b37a838552aad"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "merged_id": "test_1_2"
                    },
                    "test_null.sto.gz:md5,e3f8714af2c2c9a2b18a8e142ff85a66"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "merged_id": "test_1_2"
                    },
                    "test_null.fasta.gz:md5,bc97ed94ff93c7dc09e3163f241e649e"
                ]
            ],
            [
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:HMMER_HMMALIGN": {
                        "hmmer": 3.4
                    }
                },
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:HMMER_HMMBUILD": {
                        "hmmer": 3.4
                    }
                },
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:FILTER_RECRUITED": {
                        "python": "3.13.1",
                        "biopython": 1.85
                    }
                },
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:HMMER_HMMSEARCH": {
                        "hmmer": 3.4
                    }
                },
                {
                    "MERGE_FAMILIES:POOL_SIMILAR_COMPONENTS": {
                        "python": "3.13.1",
                        "pandas": "2.3.3",
                        "networkx": 3.5
                    }
                },
                {
                    "MERGE_FAMILIES:MERGE_SEEDS": {
                        "python": "3.13.1"
                    }
                },
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:CLIPKIT": {
                        "clipkit": "2.4.1"
                    }
                },
                {
                    "MERGE_FAMILIES:GENERATE_FAMILIES:ALIGN_SEQUENCES:FAMSA_ALIGN": {
                        "famsa": "2.4.1-45c9b2b (2025-05-09)"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-09T11:59:22.313724388"
    }
}